### PR TITLE
Remove deprecation `:reporting` project

### PR DESCRIPTION
**Pull request by lacasseio**
### Context
See issue: https://github.com/gradle/gradle/issues/1744

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [x] Verify documentation including proper use of `@since` and `@Incubating` annotations for all public APIs
- [x] Recognize contributor in release notes

### CODE REVIEW

**Comment by lacasseio** [comment]
Stage 4 CI: https://builds.gradle.org/viewLog.html?buildId=2663361&tab=buildResultsDiv&buildTypeId=Gradle_Branches_Checkpoints_Stage4FullCoverage

File: subprojects/reporting/src/main/java/org/gradle/api/reporting/ConfigurableReport.java [hunk 1]
```diff
@@ -48,18 +48,6 @@
...
      * Sets the destination for the report.
      *
-     * The file parameter is evaluated as per {@link org.gradle.api.Project#file(Object)}.
-     *
-     * @param file The destination for the report.
-     * @see #getDestination()
-     * @deprecated Replaced by {@link #setDestination(java.io.File)} and {@link #setDestination(org.gradle.api.provider.Provider)}.
-     */
-    @Deprecated
-    void setDestination(Object file);
```
Comment (bmuschko) [inline subprojects/reporting/src/main/java/org/gradle/api/reporting/ConfigurableReport.java]: We should think about removing the overridden method with the same signature on

- `SimpleReport`
- `TaskGeneratedSingleDirectoryReport`
- `TaskGeneratedSingleFileReport`

These classes are internal but we have an opportunity to get rid of the method completely. Have a look at the uses of the method in other context and see if it makes sense.

**Comment by lacasseio** [comment]
@bmuschko Should we remove [`SimpleReport#setDestination(Object)`](https://github.com/gradle/gradle/blob/master/subprojects/reporting/src/main/java/org/gradle/api/reporting/internal/SimpleReport.java#L74-L81) given [a test was explicitely written](https://github.com/gradle/gradle/blob/7a7c049a996711e264112a357970b6904800d17e/subprojects/plugins/src/test/groovy/org/gradle/api/reporting/internal/TaskGeneratedReportTest.groovy#L30-L41) for it? The other `setDestination(Object)` were removed as it can use the `super` class implementation.

**Comment by bmuschko** [comment]
@lacasseio Can you have a look if the functionality is already covered somewhere else by a test? If it isn't then I'd keep it in an assign a `File` type for setDestination.

**Comment by lacasseio** [comment]
Sorry @bmuschko. I'm reading your comment multiple time over and can't figure out the last part. The `setDestination` have overload for `Object`, `File` and `Provider<File>`. The `Object` overload simply set a `Provider<File>` and call `FileResolver#resolve` on the object. Given I can't find any usage in our code base, we could simply force developers to use the `Provider<File>` overload with `Project#file`.

**Comment by bmuschko** [comment]
@lacasseio Sorry, I misread your comment. I thought you were asking about removing the test. I'd keep the test, change the usage of `setDestination` in the test (assign a `File` instance) and remove setDestination(Object) on `SimpleReport` (which is an internal class anyway). We'll see if some other code uses in Gradle core code base when running tests.

**Comment by lacasseio** [comment]
@bmuschko I removed the test completely as it was testing that a `String` resolve to a `File`. The overloaded method was actually used in quite a lot of place. I adjust them all to pass a `File` instance. A CI run was started to ensure we aren't missing any call to the removed method.

**Comment by lacasseio** [comment]
I also rebased the branch.

**Comment by bmuschko** [comment]
Thanks for updating the PR. I just looked through it again. As discussed let's not remove the method for now as we didn't go through a full deprecation cycle yet. Let's remove it with 5.0 or potentially a bit earlier. I am going to close the PR.

**Pull request closed without merging**